MTV-1722: AIO buffer performance

anarnold97 opened the pull request:
### JIRA

* [MTV-1722](https://issues.redhat.com/browse/MTV-1722)

### PREVIEW

* [Increasing asynchronous I/O (AIO) sizes and buffer counts for NBD transport mode](https://anarnold97.github.io/MTA-Preview/MTV/aio-buffer.html#mtv-aio-buffer_mtv)

dgur commented:
I think 9.11.1 section needs to be recompiled to have all in one better wording

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,323 @@
+
+* No significant impact was observed when using AIO buffer settings with a single VM.
+
+[id="mtv-aio-buffer-environment_{context}"]
+=== Environment
+
+* The RDU2 scale lab utilizes metal OpenShift clusters and VMware vSphere servers, which employ dedicated hardware, isolated networking, and storage to eliminate external variables and guarantee consistent performance metrics.
+** *Vmware information*:
+*** vSphere Version: 7.0.3
+*** ESXi Version: 7.0.3
+*** DataStore : ISCSI & FC Luns, attached to netapp storage
+** *MTV information*:
+*** VDDK Version: 7.0.3
+*** MTV Version: 2.7.4
+*** Max VM Inflight: 20
+** *VM Specification*:
+*** *Network*: vSwitch with MTU 9000, 10 GbE ESXi interfaces.
+*** *Disks*: Single disk, 50GB size, 70% utilization.
+
```
```suggestion
```

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,307 @@
+
+If needed, ensure the changes take effect by restarting the forklift-controller pod.
+
+[id="mtv-aio-buffer-comparison_{context}"]
+== AIO buffer comparisons
+
+[id="mtv-aio-buffer-comparison-single-vm_{context}"]
+=== One VM Migration with AIO (16/4) and without AIO
+
+.One VM Migration with AIO and without AIO
+[width="99%",cols="20%,12%,20%,16%,16%,16%",options="header",]
+|===
+|AIO Buffer values |MTV |VDDK |max_vm inflight |Migration type |Total
+duration
+|N/A
+|2.7.4
+|7.0.3
+|20
+|cold
+|00:08:12
```
```suggestion
```

dgur commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,253 @@
+[id="mtv-aio-performance-results_{context}"]
+== Performance Testing Results
+
+There were several extensive tests, which were conducted across multiple configurations:
```
I don think we need to specify how we tested exactly  - it is our problem

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,252 @@
+=== Key Findings
+
+The highest migration performance was achieved with 10 VMs on a single ESXi host with the following values:
+
+* `VixDiskLib.nfcAio.Session.BufSizeIn64KB=16`
+* `vixDiskLib.nfcAio.Session.BufCount=4`
+
+
+[id="mtv-aio-buffer-improvement_{context}"]
+=== Improvement:
```
```suggestion
[id="mtv-aio-buffer-improvement_{context}"]
== Improvement:
```

dgur commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,248 @@
+[id="mtv-aio-buffer-key-findings_{context}"]
+=== Key findings
+
+The best migration performance was achieved by migrating using Multiple 10 VMs(10) on a single ESXi host with the following values:
```
Multiple VMs (10)

dgur replied:
no need 10 before VMs

dgur commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,248 @@
+* Migration time was reduced by *31.1%*, from `0:24:32` to `0:16:54`.
+* Read rate was increased from `347.83MB/s` to `504.93MB/s`.
+
+There was no significant improvement observed when using AIO buffer settings with a single VM.
```
Can you somehow bullet it and the below?

dgur commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,248 @@
+* `VixDiskLib.nfcAio.Session.BufSizeIn64KB=16`
+* `vixDiskLib.nfcAio.Session.BufCount=4`
+
+The following improvements were noted when using AIO buffer (Asynchronous Buffer Counts) settings:
```
I suggest  indent right as it relates to the above Many VMs on single host

anarnold97 replied:
@dgur - changes made and preview manually updated (Netlify has ended the license). thanks

dgur replied:
looks good now.
I need now that Tzahi and Mordechai validate the commands to set and remove it

TzahiAshkenazi commented:
need to edit section 9.11.2 
since the forklift controller name prefix is dynamic  , and on the current example its hard coded 
so please change  it to this : 

`Validating Controller Pod support for AIO values

Ensure that the forklift-controller pod in the openshift-mtv namespace supports the AIO buffer values.
Since the pod name prefix is dynamic , check the pod name first  by running the folowing command: 

`oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'`
output : forklift-controller-667f57c8f8-qllnx
This is the pod name prefix from the example : forklift-controller-667f57c8f8-qllnx

Check the environment variables of the pod by running:
`oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml``

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,246 @@
++
+[source,termina]
+----
+virt_v2v_extra_conf_config_map: "`perf`"
```
@anarnold97 
please change the following line with this one : 
virt_v2v_extra_conf_config_map: "perf"

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,246 @@
+* There was no significant improvement observed when using AIO buffer settings with multiple VMs from multiple hosts.
+
+[id="mtv-enable-aio-buffer_{context}"]
+== Enabling AIO buffer configuration
```
need to edit section 9.11.2
since the forklift controller name prefix is dynamic , and on the current example its hard coded
so please change it to this :

`Validating Controller Pod support for AIO values

Ensure that the forklift-controller pod in the openshift-mtv namespace supports the AIO buffer values.
Since the pod name prefix is dynamic , check the pod name first by running the folowing command:

oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
output : forklift-controller-667f57c8f8-qllnx
This is the pod name prefix from the example : forklift-controller-667f57c8f8-qllnx

Check the environment variables of the pod by running:
`oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml``

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,246 @@
+
+.Validating Controller Pod support for AIO values
+
+* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values. Check the environment variables of the pod by running:
```
```suggestion
* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values.
+
Since the pod name prefix is dynamic, check the pod name first by running the following command:
```

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,246 @@
+.Validating Controller Pod support for AIO values
+
+* Ensure that the `forklift-controller` pod in the `openshift-mtv` namespace supports the AIO buffer values. Check the environment variables of the pod by running:
+
```
```suggestion

+
[source,terminal]
----
oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
----

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,253 @@
+----
+oc get pods -n openshift-mtv | grep forklift-controller | awk '{print $1}'
+----
++
```
```suggestion
+
The example output is as follows:
+
[source,termina]
----
forklift-controller-667f57c8f8-qllnx
----
+
[NOTE]
====
This is the pod name prefix from the example: `forklift-controller-667f57c8f8-qllnx`
====

* Check the environment variables of the pod by running:
+
[source,termina]
----
oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
----
+
```

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,270 @@
++
+[source,termina]
+----
+virt_v2v_extra_conf_config_map: "`perf`"
```
```suggestion
virt_v2v_extra_conf_config_map: "perf"
```

anarnold97 commented:
@dgur && @TzahiAshkenazi - thanks for the reviews lads

i think i have captured the changes you requested. Please could you do *second-eyes*

Thanks

dgur commented:
It is **important** to Mention that those Settings only work with Cold Migration.

dgur commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,270 @@
+[id="mtv-aio-buffer_{context}"]
+= AIO buffer counts for customer usage
+
+This document is intended to summarize the performance team's testing on the impact of changes to the AIO buffer count, specifically to assess any effects on datastore read rates and overall migration time when using the {project-first} product.
```
The gdoc was to summarize testing - in the official docs we just need to explain what to do and why - I believe.
Let's write something like this

This chapter describes how to change NBD transport  NFC parameters for increased migration performance.

dgur commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,270 @@
+[id="mtv-aio-performance-results_{context}"]
+== Performance Testing Results
+
+There were several extensive tests, which were conducted across multiple configurations:
```
Remove. We are not doing test report

dgur commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,270 @@
+
+There were several extensive tests, which were conducted across multiple configurations:
+
+* Single VMware ESXi server compared to multiple ESXi servers.
```
remove

dgur commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,270 @@
+There were several extensive tests, which were conducted across multiple configurations:
+
+* Single VMware ESXi server compared to multiple ESXi servers.
+* Single virtual machine (VM) compared to multiple VMs.
```
remove

dgur commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,270 @@
+Adding a high log level will degrade performance and is for debugging purposes only.
+====
+
+.Disabling AIO Buffer Configuration
```
Disable Part shouldn’t  be under Unable but another paragraph

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,270 @@
+
+:_content-type: CONCEPT
+[id="mtv-aio-buffer_{context}"]
+= AIO buffer counts for customer usage
```
```suggestion
= Increasing asynchronous I/O (AIO) sizes and buffer counts for NBD transport mode
```
Sorry, but that title is too long, so i will abbreviate

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,270 @@
+[id="mtv-aio-buffer_{context}"]
+= Increasing asynchronous I/O (AIO) sizes and buffer counts for NBD transport mode
+
+This document is intended to summarize the performance team's testing on the impact of changes to the AIO buffer count, specifically to assess any effects on datastore read rates and overall migration time when using the {project-first} product.
```
```suggestion
This document describes how to change NBD transport NFC parameters for increased migration performance when using the {project-first} product.
```

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,262 @@
++
+[source,termina]
+----
+virt_v2v_extra_conf_config_map: "`perf`"
```
```suggestion
virt_v2v_extra_conf_config_map: "perf"
```

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,270 @@
+
+:_content-type: CONCEPT
+[id="mtv-aio-buffer_{context}"]
+= AIO buffer counts for customer usage
```
Apologies but this title is a bit too long and could create problems when building the docs

So as a compromise could we abbreviate to **Increasing asynchronous I/O (AIO) sizes and buffer counts for NBD transport mode**

anarnold97 commented:
@dgur & @TzahiAshkenazi - thanks for the reviews

i have added the changes, where possible, and made compromises where not.

i have updated the preview

please can i ask you to do another review?

thanks

anarnold97 commented:
@mnecas && @fabiand - Please can I ask for your comments on this PR? I think we are currently saying **Not to Publish**.

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,263 @@
++
+[source,termina]
+----
+oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
```
remove this line please  , its using the wrong pod prefix name

anarnold97 replied:
@TzahiAshkenazi - Please can you confirm the line number to be removed

TzahiAshkenazi replied:
this comment was related to previous  line you already removed with the second pod prefix name 
so according to the new PREVIEW no need to remove

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,263 @@
+oc edit forkliftcontroller -n openshift-mtv`
+----
+
+* For example, to add the following values:
+** virt_v2v_extra_args:
++
+[source,termina]
+----
+--vddk-config /mnt/extra-v2v-conf/input.conf
+----
++
+[source,termina]
+----
+virt_v2v_extra_conf_config_map: "perf"
+----
```
```suggestion

```

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,251 @@
++
+[source,termina]
+----
+oc edit forkliftcontroller -n openshift-mtv`
```
```suggestion
virt_v2v_extra_args: --vddk-config /mnt/extra-v2v-conf/input.conf
virt_v2v_extra_conf_config_map: "perf"
```

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,252 @@
+----
+oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
+----
++
+[source,terminal]
+----
+oc get pod forklift-controller-5b45f66949-vgzhx -n openshift-mtv -o yaml
```
```suggestion
```

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,248 @@
+----
+oc get pod forklift-controller-667f57c8f8-qllnx -n openshift-mtv -o yaml
+----
+----
```
```suggestion
```

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,247 @@
+= Increasing asynchronous I/O (AIO) sizes and buffer counts for NBD transport mode
+
+This document describes how to change NBD transport NFC parameters for increased migration performance when using the {project-first} product.
+
```
```suggestion

[WARNING]
.Cold Migration Only
====
Using AIO buffering is only suitable for Cold Migration use cases. 
====

```

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,253 @@
+`oc delete cm perf -n openshift-mtv`
+
+....
+  3\.	Restart the Forklift Controller Pod (Optional):
```
from this line please remove the "3\."
  3\.	Restart the Forklift Controller Pod (Optional):

need to be 
Restart the Forklift Controller Pod (Optional):

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,253 @@
+vSphere 6.7 Update 3 and later.
+Fully supported in vSphere 7.x versions, including 7.0.x.
+
+|VDDK Version
```
this section missing  the next recommendation for VDDK : 

VDDK Version:
Requires VDDK 6.7.3 or newer. The VDDK version typically aligns with the ESXi version installed on the vSphere host.

TzahiAshkenazi replied:
me and daniel made some changes to this section please use the update one : 

VDDK /vSpheres Versions: 

Tested with : 
vSphere : 7.0.3
VDDK : 7.0.3 
* for other  VDDK / vSpheres versions please check the AIO buffer support in the official VMware documentation

anarnold97 commented:
@dgur & @TzahiAshkenazi - thanks for the comments. Changes made and the preview updated

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,261 @@
++
+[source,termina]
+----
+oc -n openshift-mtv create cm perf`
```
replace this line: oc -n openshift-mtv create cm perf`
with this line: oc -n openshift-mtv create cm perf

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,261 @@
++
+[source,termina]
+----
+`cat /mnt/extra-v2v-conf/input.conf`
```
replace  this line:  `cat /mnt/extra-v2v-conf/input.conf`
with this one:  cat /mnt/extra-v2v-conf/input.conf

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,261 @@
+[id="mtv-aio-buffer-key-requirements_{context}"]
+== Key requirements for AIO Buffer (Asynchronous Buffer Counts) support
+
+.Key requirements for AIO Buffer support
```
replace this title: 9.11.4. Key requirements for AIO Buffer (Asynchronous Buffer Counts) support 

with 9.11.4  VDDK /vSpheres Versions:

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,261 @@
+== Key requirements for AIO Buffer (Asynchronous Buffer Counts) support
+
+.Key requirements for AIO Buffer support
+[width="100%",cols="54%,46%",options="header",]
```
replace lines 243-261 
with this new content : 

Tested with : 
       - vSphere : 7.0.3
       - VDDK : 7.0.3 
         * for other  VDDK / vSpheres versions please check the AIO buffer support in the official VMware 
            documentation

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,261 @@
++
+[source,termina]
+----
+echo -e "`VixDiskLib.nfcAio.Session.BufSizeIn64KB=16\nvixDiskLib.nfcAio.Session.BufCount=4`" | base64`
```
replace this old line: 
 echo -e "`VixDiskLib.nfcAio.Session.BufSizeIn64KB=16\nvixDiskLib.nfcAio.Session.BufCount=4`" | base64`
 
with this  line:
echo -e "VixDiskLib.nfcAio.Session.BufSizeIn64KB=16\nvixDiskLib.nfcAio.Session.BufCount=4" | base64

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,261 @@
++
+[source,termina]
+----
+virt_v2v_extra_args: --vddk-config /mnt/extra-v2v-conf/input.conf
```
change lines 88+89 
with those  new  lines content 
virt_v2v_extra_args: "--vddk-config /mnt/extra-v2v-conf/input.conf"
virt_v2v_extra_conf_config_map: "perf"

MrQuality commented:
@anarnold97 hi, Can you please update about the changes @TzahiAshkenazi asked to change?

anarnold97 commented:
@MrQuality - please be aware that this is not the only item in my work queue. However, i have updated this PR and the preview

@dgur - FYI

dgur commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,246 @@
+[WARNING]
+.Cold Migration Only
+====
+Using AIO buffering is only suitable for Cold Migration use cases.
```
We saw  that if AIO is configured and we try warm warm will fail...
Should we also add something ike "Desbale AOI settings before initilazing Warm Migration - Refer  chapter 9.11.3. "Disabling AIO Buffer Configuration"

anarnold97 commented:
PR and preview updated

dgur commented:
Reviewed Looks good for me

anarnold97 commented:
@TzahiAshkenazi & @dgur - updated and preview updated

TzahiAshkenazi commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,250 @@
+.Editing ForkliftController Configuration
+
+* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values.
++
```
add below line 84 , the command to edit the ForkliftController as  follow : 
run the following command to edit the ForkliftController object :
oc edit forkliftcontroller -n openshift-mtv

TzahiAshkenazi commented:
@anarnold97 
i have added missing command for section : 
Editing ForkliftController Configuration
the missing line was under line 84 which need to include the oc command to edit the forklift-controller 
@dgur F.Y.Y
the last section  number 9.11.4
is different from what we agreed on the main aio buffer document 
do we want to change this section or to stay with the current one ?

dgur commented:
I am ok with what Andy wrote, it is clear that there are required versions,
and we tested only with some. Others they need to test

On Sun, 15 Dec 2024, 10:48 Tzahi Ashkenazi, ***@***.***>
wrote:

> @anarnold97 <https://github.com/anarnold97>
> i have added missing command for section :
> Editing ForkliftController Configuration
> the missing line was under line 84 which need to include the oc command to
> edit the forklift-controller
> @dgur <https://github.com/dgur> F.Y.Y
> the last section number 9.11.4
> is different from what we agreed on the main aio buffer document
> do we want to change this section or to stay with the current one ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubev2v/forklift-documentation/pull/598#issuecomment-2543562111>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGJQIUPUZRYOGGQSWP6BS32FU66TAVCNFSM6AAAAABTBHQAN6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNBTGU3DEMJRGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

anarnold97 commented:
@dgur - clearly this is not going to be approved. After numerous reviews i am closing it and not continuing.

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,250 @@
+
+.Editing ForkliftController Configuration
+
+* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values.
```
```suggestion
* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values, by running the following command:
++
[source,termina]
----
oc edit forkliftcontroller -n openshift-mtv
----
```

anarnold97 commented on documentation/modules/mtv-aio-buffer.adoc:
```diff
@@ -0,0 +1,255 @@
+.Editing ForkliftController Configuration
+
+* In the `openshift-mtv namespace`, edit the `ForkliftController` object to include the AIO buffer values, by running the following command:
+++
```
```suggestion
+
```

dgur commented:
Looks good for me, with the last fix from today, Thank you Andy! 
Can you also generate  the updated preview, and publush?
